Sanitize HTML for table and publication descriptions

alisonrclarke opened the pull request.

alisonrclarke committed:
Set table name with text method not html to avoid issues with < and > in name

alisonrclarke committed:
Sanitize HTML in submission abstract and table descriptions

We want to escape rather than strip disallowed tags so we use bleach
rather than invenio-formatter.
We have to take a guess about how to deal with text that is not HTML but
contains < or > - this works for the test cases found but we should add
tips in the hepdata-submission docs for what to do if the HTML output is
not as expected.

alisonrclarke committed:
Merge master into fix/sanitize-table-html

alisonrclarke committed:
Removed headings from list of allowed tags

alisonrclarke committed:
Added test case for < with spaces.

alisonrclarke committed:
Added note to update hepdata-submission docs if changing allowed tags etc

GraemeWatt commented on hepdata/config.py:
```diff
@@ -300,6 +300,11 @@ def _(x):
+# HTML attributes and tags allowed in record descriptions
+# If these are modified, update tips.rst in hepdata-submission
+ALLOWED_HTML_ATTRS = {'*': ['class'], 'a': ['href', 'title', 'name', 'class', 'rel'], 'abbr': ['title'], 'acronym': ['title']}
+ALLOWED_HTML_TAGS = ['a', 'abbr', 'acronym', 'b', 'blockquote', 'br', 'code', 'div', 'em', 'i', 'li', 'ol', 'p', 'pre', 'span', 'strike', 'strong', 'sub', 'sup', 'u', 'ul']
```
I know these attributes and tags were copied from Invenio-Formatter, but do you see a use case for the `class` attribute?  I suppose an Uploader could use a class name given in a HEPData style sheet to achieve the same styling, but then we'd need to document possible class names for this to be useful.  I'm wondering whether it would be simpler to remove `class` as a possible attribute and update `tips.rst` in the hepdata-submission docs.

GraemeWatt commented on hepdata/modules/records/templates/hepdata_records/components/table_list.html:
```diff
@@ -124,7 +124,7 @@ <h4 itemprop="name">{{ table.name }}</h4>
 
                     {% if table.description != '.' %}
-                        <p itemprop="description">{{ table.description }}</p>
+                        <p itemprop="description">{{ table.description|safe }}</p>
```
Maybe it'd be better to escape HTML in the list of tables (i.e. remove `|safe`) as was the previous behaviour.  The list of tables is mainly used by the user to select the table of interest.  If the text contains an HTML link, maybe the user would click the link accidentally instead of selecting the table.  The use of `sanitize_html` can then be removed from `process_data_tables` above.

alisonrclarke replied:
I see what you mean but I suggest we use `sanitize_html` with no tags allowed and `strip=True` so that we just remove any HTML elements and leave the plain text, rather than have the table list littered with `&lt;` etc.

GraemeWatt commented on hepdata/utils/miscellaneous.py:
```diff
@@ -20,9 +20,52 @@
+
+    # The lines below were copied from invenio-formatter but with the addition
+    # of the strip parameter.
+    value = value.strip()
```
The use of `strip` gives an exception if `value` is `None`, for example, for a newly created record without a data abstract.  I guess you just need to add something like:
```python
if value is None:
    return value
```

GraemeWatt commented on tests/miscellaneous_test.py:
```diff
@@ -62,3 +62,43 @@ def test_file_extractor(app):
+def test_sanitize_html(app):
+    with app.app_context():
+        test_cases = [
+            ("<b>Here is some bold text</b> and <script>here is a dodgy script</script>",
```
Please add the test case `None` where the sanitized result is also `None`.

GraemeWatt reviewed (changes requested):
Thanks, this looks good.  Please address the minor comments before I merge.

alisonrclarke committed:
Remove 'class' from list of allowed attributes.

alisonrclarke committed:
Modified sanitize_html behaviour

In order to allow strip to work in more cases, replaced the length
checking with a specific test for strings like 1<x<2 which cause
bleach to remove things after the first tag. These are escaped first
before calling bleach. It also now checks for Nones.

Also added more test cases.

alisonrclarke committed:
Strip all tags for table descriptions used in list.

alisonrclarke committed:
Merge branch 'master' into fix/sanitize-table-html
